### ISSUE TITLE
"dateFile" not working on Windows server, and did not create a new log file every day

### ISSUE THREAD

**Issue by Vidtion** [log4js-node/log4js-node#909]
The file `default.log`  never recreate.  The size of the file keeps growing every day, so is the daily *gz files. The `default.log` contains the content from 2019/06/20 to 2019/07/03.

```
2019/06/20  00:00         2,155,220 default.2019-06-19.log.gz
2019/06/21  00:00         5,251,746 default.2019-06-20.log.gz
2019/06/22  00:01        11,917,136 default.2019-06-21.log.gz
2019/06/23  00:00        15,745,102 default.2019-06-22.log.gz
2019/06/24  00:00        22,593,144 default.2019-06-23.log.gz
2019/06/25  00:00        30,018,988 default.2019-06-24.log.gz
2019/06/26  00:00        40,214,672 default.2019-06-25.log.gz
2019/06/27  00:01        50,778,637 default.2019-06-26.log.gz
2019/06/28  00:00        66,371,073 default.2019-06-27.log.gz
2019/06/29  00:00        79,911,509 default.2019-06-28.log.gz
2019/06/30  00:00        93,948,478 default.2019-06-29.log.gz
2019/07/01  00:00       127,523,410 default.2019-06-30.log.gz
2019/07/02  00:00       130,904,256 default.2019-07-01.log.gz
2019/07/03  00:01       135,744,068 default.2019-07-02.log.gz
2019/07/03  08:07     1,188,977,973 default.log
```

`"log4js": "^4.3.1"`

```
var path = require('path');
let log4js = require('log4js');
let logPath = path.join(__dirname, '../log/default.log');

log4js.configure({
    appenders: {
        expressApp: {
            daysToKeep:30,
            type: 'dateFile',
            filename: logPath,
            keepFileExt:true,
            compress:true,
        }
    },
    categories: {
        default: {
            appenders: [ 'expressApp' ],
            level: 'all'
        }
    }
});
```

**Comment by nomiddlename** [comment]
Looks like it is copying the file every night, but not truncating the default.log file afterwards. I'll see if I can reproduce the problem and work on a fix.

**Comment by Vidtion** [comment]
> Looks like it is copying the file every night, but not truncating the default.log file afterwards. I'll see if I can reproduce the problem and work on a fix.

I add `pattern: '.yyyy-MM-dd'` to the appender, and the problem is gone.

**Comment by nomiddlename** [comment]
Good to know your problem is fixed, looks like the default pattern is not being passed to the underlying library. That's an easier thing to fix anyway :)

**Comment by Vidtion** [comment]
> Good to know your problem is fixed, looks like the default pattern is not being passed to the underlying library. That's an easier thing to fix anyway :)

The problem shows again, the  pattern: '.yyyy-MM-dd' didn't help since the third day. Maybe its not the problem about the pattern. Maybe its because my log file still very busy at 00:00 of the day.

**Comment by nomiddlename** [comment]
I fixed a couple of weird bugs in the file rolling code in log4js@4.5.1 - could you upgrade and see if it makes any difference please?

**Comment by damienwhaley** [comment]
I think we have hit the same thing. We're on Mac/Linux.

We have the config being:

```
{
  type: 'dateFile',
  filename: '/var/log/logfile.log',
  pattern: '_yyyy-MM-dd',
  alwaysIncludePattern: false,
  compress: true,
  layout: {
    type: 'basic',
  },
}
```

When it used to roll over the files the file names were:

```
logfile.log # Current Log File
logfile.log_2019-04-09.gz
logfile.log_2019-04-11.gz
logfile.log_2019-04-30.gz
```

But now when it rolls over the files it does:

```
logfile.log # Current Log File
logfile.log._2019-04-09.gz
logfile.log._2019-04-11.gz
logfile.log._2019-04-30.gz
```

It seems that it adds an extra `.` in front of the pattern if it is a custom pattern.

On our servers we have the log file filenames being the original ones, and when it tries to roll a log file we see the following exception:

```
Error: String '2019-04-07' could not be parsed as '_yyyy-MM-dd'
    at extractDateParts (/var/app/application/package/node_modules/date-format/lib/index.js:114:9)
    at Function.parse (/var/app/application/package/node_modules/date-format/lib/index.js:122:10)
    at _.sortBy.n (/var/app/application/package/node_modules/streamroller/lib/RollingFileWriteStream.js:203:24)
    at /var/app/application/package/node_modules/lodash/lodash.js:3727:18
    at arrayMap (/var/app/application/package/node_modules/lodash/lodash.js:639:23)
    at /var/app/application/package/node_modules/lodash/lodash.js:3726:24
    at /var/app/application/package/node_modules/lodash/lodash.js:3543:27
    at /var/app/application/package/node_modules/lodash/lodash.js:4881:15
    at baseMap (/var/app/application/package/node_modules/lodash/lodash.js:3542:7)
    at baseOrderBy (/var/app/application/package/node_modules/lodash/lodash.js:3725:20)
```

If we rename the existing rolled-over log files to the new filename (with the additional `.`) then it works.

I've had a quick look through the code and it looks like the exception is being thrown when it is going through the existing log files.

Note that we have also upgraded from `3.0.5` to `4.5.1`. However I believe the log rolling this was working in `4.1.1`.

I hope this helps!

**Comment by damienwhaley** [comment]
Update - it looks like the 4.x releases all have the same problem. We think it is a change in behaviour of the `streamroller` dependency.

**Comment by nomiddlename** [comment]
I think the upgrade from 3.x to 4.x probably introduced the extra `.` in the filenames. `4.5.1` introduced the problem of trying to parse the older file names and getting an error. As a workaround, you could downgrade your log4js version (and gain a different set of bugs maybe). I'm working on a fix for this and a couple of other log-rolling bugs at the moment, but I can't give you a timeline on when I'll be done.

**Comment by damienwhaley** [comment]
Thanks! We've pinned our dependency to `3.0.6` and it's back to normal (if there are any bugs we have not found them). We'll keep an eye out for future releases and try to upgrade again when they come out.

**Comment by nomiddlename** [comment]
log4js@5.1.0 should fix some of these problems (particularly the error trying to parse old filenames - it will now just ignore filename it cannot parse).

**Comment by nomiddlename** [comment]
The problem with the file not being truncated should be fixed in log4js@5.2.0 - there was a bug if the file was busy when we tried to rotate it.

**Comment by markrad** [comment]
Adding to this issue, on Windows it appears that if you do not have ```alwaysIncludePattern: true``` on a restart it will smash the current log. When it is true it will append instead as expected.

**Comment by nomiddlename** [comment]
@markrad that problem should be fixed in log4js 5.2.2

**Comment by nomiddlename** [comment]
I'm going to close this issue, because I've had confirmation from another user that a related problem is fixed. If you're still seeing this issue with log4js > 5.2.2 then please re-open.
